crowdsourcehinter/crowdsourcehinter.py

```diff
@@ -0,0 +1,352 @@
+        Args:
+          answer: This is equal to answer from get_hint, the answer the student submitted
+        Returns 0 if no hints to show exist
+        """
+        isreported = []
+        self.incorrect_answers.append(str(answer))
+        if str(answer) not in self.hint_database:
+            # add incorrect answer to hint_database if no precedent exists
+            self.hint_database[str(answer)] = {}
+            return str(0)
+        for hint_keys in self.hint_database[str(answer)]:
+            for reported_keys in self.reported_hints:
+                if hint_keys == reported_keys:
+                    isreported.append(hint_keys)
+        if (len(self.hint_database[str(answer)]) - len(isreported)) > 0:
+            return str(1)
+        else:
+            return str(0)
+
+    @XBlock.json_handler
+    def get_used_hint_answer_data(self, data, suffix=''):
+        """
+        This function helps to facilitate student rating of hints and contribution of new hints.
+        Specifically this function is used to send necessary data to JS about incorrect answer
+        submissions and hints. It also will return hints that have been reported, although this
+        is only for Staff.
+        Returns:
+          used_hint_answer_text: This dicitonary contains reported hints/answers (if the user is staff) and the
+                         first hint/answer pair that the student submitted for a problem.
+        """
+        # used_hint_answer_text is a dictionary of hints (or lack thereof) used for a
+        # specific answer, as well as 2 other random hints that exist for each answer
+        # that were not used. The keys are the used hints, the values are the
+        # corresponding incorrect answer
+        used_hint_answer_text = {}
+        if self.get_user_is_staff():
+            for key in self.reported_hints:
+                used_hint_answer_text[key] = str("Reported")
+        if len(self.incorrect_answers) == 0:
+            return used_hint_answer_text
+        else:
+            for index in range(0, len(self.used)):
+                # each index is a hint that was used, in order of usage
+                if str(self.used[index]) in self.hint_database[self.incorrect_answers[index]]:
+                    # add new key (hint) to used_hint_answer_text with a value (incorrect answer)
+                    used_hint_answer_text[str(self.used[index])] = str(self.incorrect_answers[index])
+                    self.incorrect_answers = []
+                    self.used = []
+                    return used_hint_answer_text
+                else:
+                    # if the student's answer had no hints (or all the hints were reported and unavailable) return None
+                    used_hint_answer_text[None] = str(self.incorrect_answers[index])
+                    self.incorrect_answers = []
+                    self.used = []
+                    return used_hint_answer_text
+        self.incorrect_answers = []
```

Comment:
Minor: this line is nearly unreachable: it will only execute if self.used is empty.  the logic in this function can be simplified.  Notice that three different places end the function with the same three lines of code.